include/nlohmann/detail/iterators/internal_iterator.hpp

```diff
@@ -1,25 +1,189 @@
 #pragma once
 
+#include <cassert>
+#include <cstdint>
+#include <type_traits>
+
 #include <nlohmann/detail/iterators/primitive_iterator.hpp>
+#include <nlohmann/detail/meta/memory.hpp>
 
 namespace nlohmann
 {
 namespace detail
 {
+
+template<class BasicJsonType>
+class internal_iterator_base
+{
+  public:
+    using ObjectIterator = typename BasicJsonType::object_t::iterator;
+    using ArrayIterator = typename BasicJsonType::array_t::iterator;
+
+    ObjectIterator& object_iterator()
+    {
+        return *reinterpret_cast<ObjectIterator*>(&m_data);
+    }
+    ObjectIterator object_iterator() const
+    {
+        return *reinterpret_cast<const ObjectIterator*>(&m_data);
+    }
+
+    ArrayIterator& array_iterator()
+    {
+        return *reinterpret_cast<ArrayIterator*>(&m_data);
+    }
+    ArrayIterator array_iterator() const
+    {
+        return *reinterpret_cast<const ArrayIterator*>(&m_data);
+    }
+
+    primitive_iterator_t& primitive_iterator()
+    {
+        return *reinterpret_cast<primitive_iterator_t*>(&m_data);
+    }
+    primitive_iterator_t primitive_iterator() const
+    {
+        return *reinterpret_cast<const primitive_iterator_t*>(&m_data);
+    }
+
+  private:
+    // take the maximum size of all three iterators
+    static constexpr std::size_t kMaxIteratorSize = sizeof(ObjectIterator) < sizeof(ArrayIterator) ? (sizeof(ArrayIterator) < sizeof(primitive_iterator_t)
+            ? sizeof(primitive_iterator_t)
+            : sizeof(ArrayIterator))
+            : (sizeof(ObjectIterator) < sizeof(primitive_iterator_t)
+               ? sizeof(primitive_iterator_t)
+               : sizeof(ObjectIterator));
+
+    // take the maximum alignment of all three iterators
+    static constexpr std::size_t kMaxIteratorAlignment = std::alignment_of<ObjectIterator>::value < std::alignment_of<ArrayIterator>::value ? (std::alignment_of<ArrayIterator>::value < std::alignment_of<primitive_iterator_t>::value
+            ? std::alignment_of<primitive_iterator_t>::value
+            : std::alignment_of<ArrayIterator>::value)
+            : (std::alignment_of<ObjectIterator>::value < std::alignment_of<primitive_iterator_t>::value
+               ? std::alignment_of<primitive_iterator_t>::value
+               : std::alignment_of<ObjectIterator>::value);
+
+  protected:
+    typename std::aligned_storage<kMaxIteratorSize, kMaxIteratorAlignment>::type m_data;
```

Comment:
Please use `alignas(kMaxIteratorAlignment) std::uint8_t m_data[kMaxIteratorSize];`

`std::aligned_storage` has been deprecated.

Reply:
PR is ready for review (Microsoft issues fixed ;)), although I still have different formatter output with `make amalgamate` (my `astyle` version is `Artistic Style Version 3.1`). Is it an issue, and if so, why do I have all these differences ? (see `include/nlohmann/json.hpp` for instance)